File: backend/BedemApi/Models/Roles.cs
using System.Security.Claims;

namespace BedemApi.Models;

/// <summary>
/// Every role the site knows about, and the role lists that [Authorize] needs.
///
/// The ladder is: a Visitor registers himself and can only talk (comments and
/// likes). A Member is someone the organisation actually let in — he votes at
/// the assembly. The three staff roles are a Member plus exactly one area they
/// administer: Moderator the news, Finance the money, Assembly the sessions.
/// Nobody but Admin holds two areas at once, so a compromised staff account
/// can only damage its own section.
/// </summary>
public static class Roles
{
    public const string Admin = "Admin";
    public const string Moderator = "Moderator";
    public const string Finance = "Finance";
    public const string Assembly = "Assembly";
    public const string Member = "Member";
    public const string Visitor = "Visitor";

    // -----------------------------------------------------------------------
    // Role lists, in the comma-separated shape [Authorize(Roles = ...)] wants.
    // Controllers read the same constants through User.IsIn(), so an attribute
    // and the check behind it can never drift apart.
    // -----------------------------------------------------------------------

    /// <summary>Write access to news. Everyone else reads what is published.</summary>
    public const string ManageNews = $"{Moderator},{Admin}";

    /// <summary>Write access to finance. Everyone else reads the public figures.</summary>
    public const string ManageFinance = $"{Finance},{Admin}";

    /// <summary>Write access to assembly sessions — agenda items and the room.</summary>
    public const string ManageAssembly = $"{Assembly},{Admin}";

    /// <summary>Inbox work: problem reports, membership applications, bot log.</summary>
    public const string ManageSubmissions = $"{Moderator},{Admin}";

    /// <summary>Comment moderation: approving pending comments and deleting them.</summary>
    public const string ManageComments = $"{Moderator},{Admin}";

    /// <summary>Accounts, roles and the audit log.</summary>
    public const string ManageUsers = Admin;

    /// <summary>Anyone who has some section of his own to administer.</summary>
    public const string Staff = $"{Moderator},{Finance},{Assembly},{Admin}";

    /// <summary>
    /// Read access to the admin panel: news drafts, the finance ledger, problem
    /// reports, membership applications. Everyone the organisation admitted may
    /// look at what the others are doing; only his own section gets buttons.
    /// A Visitor is absent — registering on the site is not membership.
    ///
    /// Wider than <see cref="AssemblyParticipants"/>: an Admin reads everything
    /// and chairs the assembly, but does not sit in it.
    /// </summary>
    public const string ViewPanel = $"{Member},{Moderator},{Finance},{Assembly},{Admin}";

    /// <summary>
    /// Who actually sits in the assembly: has a seat in the hall, counts towards
    /// the quorum, is scored for turning up, and votes.
    ///
    /// A Visitor is absent because registering on the site is not membership.
    /// An Admin is absent for a different reason: administering the site is not
    /// being a member of the association. He still runs a sitting — that is
    /// <see cref="ManageAssembly"/> — but running the room and deciding in it
    /// are separate things, and a technical account should not carry a vote.
    /// </summary>
    public const string AssemblyParticipants = $"{Member},{Moderator},{Finance},{Assembly}";

    /// <summary>
    /// The same roll as <see cref="AssemblyParticipants"/>, in the shape a query
    /// wants. Kept beside it so the two can never say different things.
    /// </summary>
    public static readonly IReadOnlyList<string> AssemblyRoll = new[]
    {
        Member, Moderator, Finance, Assembly
    };

    /// <summary>Every role, for validation when an admin changes someone's role.</summary>
    public static readonly IReadOnlyList<string> All = new[]
    {
        Admin, Moderator, Finance, Assembly, Member, Visitor
    };

    /// <summary>
    /// Roles an admin can hand out when creating an account outright. Admin is
    /// missing on purpose: promoting someone to Admin stays a separate,
    /// deliberate step through the role-change endpoint.
    /// </summary>
    public static readonly IReadOnlyList<string> Creatable = new[]
    {
        Moderator, Finance, Assembly, Member
    };

    public static bool IsKnown(string? role) => role != null && All.Contains(role);
}

public static class RoleClaimExtensions
{
    /// <summary>
    /// True when the caller holds any role in <paramref name="roleList"/> — the
    /// same comma-separated constant used by the [Authorize] attribute.
    /// </summary>
    public static bool IsIn(this ClaimsPrincipal principal, string roleList)
    {
        if (principal.Identity?.IsAuthenticated != true) return false;

        foreach (var role in roleList.Split(','))
        {
            if (principal.IsInRole(role.Trim())) return true;
        }

        return false;
    }
}

File: src/constants/roles.js
// Mirrors backend/BedemApi/Models/Roles.cs. The server is what actually enforces
// any of this — everything here only decides what the UI bothers to render, so a
// stale copy costs a 403 instead of a hole.

export const ROLES = {
  ADMIN: 'Admin',
  MODERATOR: 'Moderator',
  FINANCE: 'Finance',
  ASSEMBLY: 'Assembly',
  MEMBER: 'Member',
  VISITOR: 'Visitor',
};

export const ROLE_LABELS = {
  [ROLES.ADMIN]: 'Administrator',
  [ROLES.MODERATOR]: 'Moderator',
  [ROLES.FINANCE]: 'Finansije',
  [ROLES.ASSEMBLY]: 'Skupština',
  [ROLES.MEMBER]: 'Aktivni član',
  [ROLES.VISITOR]: 'Posetilac',
};

// What each role is for, shown next to the role picker so an admin does not
// have to keep the table in his head.
export const ROLE_DESCRIPTIONS = {
  [ROLES.ADMIN]: 'Puna prava nad sajtom, nalozima i dnevnikom izmena.',
  [ROLES.MODERATOR]: 'Aktivni član + upravlja vestima, prijavama i zahtevima za članstvo.',
  [ROLES.FINANCE]: 'Aktivni član + upravlja finansijama.',
  [ROLES.ASSEMBLY]: 'Aktivni član + saziva skupštinu i vodi dnevni red.',
  [ROLES.MEMBER]: 'Član udruženja: glasa na skupštini, komentariše i lajkuje.',
  [ROLES.VISITOR]: 'Registrovan nalog: samo komentari i lajkovi.',
};

// Roles an admin hands out when creating an account. Admin is missing on
// purpose — promoting to Admin is a separate step, same as on the server.
export const CREATABLE_ROLES = [
  ROLES.MODERATOR,
  ROLES.FINANCE,
  ROLES.ASSEMBLY,
  ROLES.MEMBER,
];

export const ALL_ROLES = [
  ROLES.ADMIN,
  ROLES.MODERATOR,
  ROLES.FINANCE,
  ROLES.ASSEMBLY,
  ROLES.MEMBER,
  ROLES.VISITOR,
];

export function roleLabel(role) {
  return ROLE_LABELS[role] || role || '';
}

const holds = (user, ...roles) => !!user && roles.includes(user.role);

export const canManageNews = (user) => holds(user, ROLES.MODERATOR, ROLES.ADMIN);
export const canManageFinance = (user) => holds(user, ROLES.FINANCE, ROLES.ADMIN);
export const canManageAssembly = (user) => holds(user, ROLES.ASSEMBLY, ROLES.ADMIN);
export const canManageSubmissions = (user) => holds(user, ROLES.MODERATOR, ROLES.ADMIN);
export const canManageComments = (user) => holds(user, ROLES.MODERATOR, ROLES.ADMIN);
export const canManageUsers = (user) => holds(user, ROLES.ADMIN);

// Who actually sits in the assembly: has a seat in the hall, counts towards the
// quorum, is scored for turning up, and votes. Mirrors Roles.AssemblyParticipants.
//
// A Visitor is absent because registering on the site is not membership. An
// Admin is absent for a different reason: he runs a sitting (canManageAssembly)
// but does not decide in it, so a technical account carries no vote.
export const canVoteInAssembly = (user) =>
  holds(user, ROLES.MEMBER, ROLES.MODERATOR, ROLES.FINANCE, ROLES.ASSEMBLY);

// Reading the panel and writing in it are separate: everyone the organisation
// admitted may look at what the others are doing, but only his own section
// gets buttons. Accounts and the audit log are the exception — those are about
// oversight of the staff itself and stay with the Admin.
export const canAccessAdmin = (user) =>
  holds(user, ROLES.MEMBER, ROLES.MODERATOR, ROLES.FINANCE, ROLES.ASSEMBLY, ROLES.ADMIN);

// Where the admin panel drops you: your own section if you have one, otherwise
// the news list, which is what a read-only visit is usually about.
export function adminLandingPath(user) {
  if (canManageAssembly(user) && !canManageNews(user)) return '/admin/skupstina';
  if (canManageFinance(user) && !canManageNews(user)) return '/admin/finance';
  if (canAccessAdmin(user)) return '/admin/news';
  return '/';
}
